builtin/mainmenu/settings/components.lua

```diff
@@ -0,0 +1,378 @@
+				local value = fields[dialog_name .. "_accepted"]
+				if value ~= nil then
+					core.settings:set(setting.name, value)
+				end
+				return true
+			end
+			if fields["set_" .. setting.name] then
+				local value = fields[setting.name]
+				if value ~= nil then
+					core.settings:set(setting.name, value)
+				end
+				return true
+			end
+		end,
+	}
+end
+
+
+function make.v3f(setting)
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = vector.from_string(core.settings:get(setting.name) or setting.default)
+			self.changed = value ~= vector.from_string(setting.default)
+
+			-- Allocate space for "Set" button
+			avail_w = avail_w - 1
+
+			local fs = "label[0,0.1;" .. get_label(setting) .. "]"
+
+			local field_width = (avail_w - 3*0.25) / 3
+
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				0, field_width, setting.name .. "_x", "X", value.x)
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				field_width + 0.25, field_width, setting.name .. "_y", "Y", value.y)
+			fs = fs .. ("field[%f,0.6;%f,0.8;%s;%s;%s]"):format(
+				2 * (field_width + 0.25), field_width, setting.name .. "_z", "Z", value.z)
+
+			fs = fs .. ("button[%f,0.6;1,0.8;%s;%s]"):format(avail_w, "set_" .. setting.name, fgettext("Set"))
+
+			return fs, 1.4
+		end,
+
+		on_submit = function(self, fields)
+			if fields["set_" .. setting.name] then
+				local x = tonumber(fields[setting.name .. "_x"])
+				local y = tonumber(fields[setting.name .. "_y"])
+				local z = tonumber(fields[setting.name .. "_z"])
+				if is_valid_number(x) and is_valid_number(y) and is_valid_number(z) then
+					core.settings:set(setting.name, vector.new(x, y, z):to_string())
+				else
+					core.log("error", "Invalid vector: " .. dump({x, y, z}))
+				end
+				return true
+			end
+		end,
+	}
+end
+
+
+function make.flags(setting)
+	local checkboxes = {}
+
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local fs = {
+				"label[0,0.1;" .. get_label(setting) .. "]",
+			}
+
+			local value = core.settings:get(setting.name) or setting.default
+			self.changed = value:gsub(" ", "") ~= setting.default:gsub(" ", "")
+
+			checkboxes = {}
+			for _, name in ipairs(value:split(",")) do
+				name = name:trim()
+				if name:sub(1, 2) == "no" then
+					checkboxes[name:sub(3)] = false
+				elseif name ~= "" then
+					checkboxes[name] = true
+				end
+			end
+
+			local columns = math.max(math.floor(avail_w / 2.5), 1)
+			local column_width = avail_w / columns
+			local x = 0
+			local y = 0.3
+
+			for _, possible in ipairs(setting.possible) do
+				if possible:sub(1, 2) ~= "no" then
+					if x >= avail_w then
+						x = 0
+						y = y + 0.5
+					end
+
+					local is_checked = checkboxes[possible]
+					fs[#fs + 1] = ("checkbox[%f,%f;%s;%s;%s]"):format(
+						x, y + 0.25, setting.name .. "_" .. possible,
```

Comment:
If you replace the `local y = 0.3` above with `local y = 0.55`, then this `y + 0.25` is unnecessary.